sqlx-core/src/net/socket/mod.rs

```diff
@@ -182,10 +183,87 @@ impl<S: Socket + ?Sized> Socket for Box<S> {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+pub struct TcpKeepalive {
+    pub time: Option<Duration>,
+    pub interval: Option<Duration>,
+    pub retries: Option<u32>,
+}
```

Comment:
If we want to encapsulate instead of using `socket2::TcpKeepalive` in the public api (that is something that @abonander should decide), we should use a [newtype wrapper](https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html) instead of just copying the other code.

```suggestion
pub struct TcpKeepalive(socket2::TcpKeepalive)
```

Reply:
That's reasonable, thx

Reply:
That's a good idea. But `socket2::TcpKeepalive` is not `Copy` (Sorry I made a mistake before).
So doing so we won't have `Copy` trait.

Reply:
Interesting. I don't know how I came up with the idea that it does. Must have misread a part of the code.

Reply:
Yeah, I have mixed feelings about using a bespoke struct here.

These builder methods could just be directly on the `ConnectOptions` structs.

I suppose being able to define `TcpKeepalive` separately in a const could be nice for reusability, but at the same time, having to import and build a separate type would also be a little annoying if you only use it once.

I think maybe because `interval` and `retries` aren't supported on all platforms, we should only expose the `tcp_keepalive_time()` on the `ConnectOptions` builders. We can always add the others later.

Reply:
These fields should also be `#[doc(hidden)]` or just private.